db/compaction/compaction_job.cc

```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
+};
+
+// offset_of is used to get the offset of a class data member
+// ex: offset_of(&ColumnFamilyDescriptor::options)
+// This call will return the offset of options in ColumnFamilyDescriptor class
+//
+// This is the same as offsetof() but allow us to work with non standard-layout
+// classes and structures
+// refs:
+// http://en.cppreference.com/w/cpp/concept/StandardLayoutType
+// https://gist.github.com/graphitemaster/494f21190bb2c63c5516
+static ColumnFamilyDescriptor dummy_cfd("", ColumnFamilyOptions());
+template <typename T1>
+int offset_of(T1 ColumnFamilyDescriptor::*member) {
+  return int(size_t(&(dummy_cfd.*member)) - size_t(&dummy_cfd));
+}
+
+static CompactionServiceInput dummy_cs_input;
+template <typename T1>
+int offset_of(T1 CompactionServiceInput::*member) {
+  return int(size_t(&(dummy_cs_input.*member)) - size_t(&dummy_cs_input));
+}
+
+static std::unordered_map<std::string, OptionTypeInfo> cfd_type_info = {
+    {"name",
+     {0, OptionType::kBinaryString, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+    {"options",
+     {offset_of(&ColumnFamilyDescriptor::options), OptionType::kConfigurable,
+      OptionVerificationType::kNormal, OptionTypeFlags::kNone,
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const std::string& value, char* addr) {
+        auto conf = CFOptionsAsConfigurable(ColumnFamilyOptions());
+        auto status = conf->ConfigureFromString(opts, value);
+        auto cf_options = reinterpret_cast<ColumnFamilyOptions*>(addr);
+        *cf_options = *conf->GetOptions<ColumnFamilyOptions>(
+            OptionsHelper::kCFOptionsName);
+        return status;
+      },
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const char* addr, std::string* value) {
+        const auto cf_options =
+            reinterpret_cast<const ColumnFamilyOptions*>(addr);
+        auto conf = CFOptionsAsConfigurable(*cf_options);
+        std::string result;
+        auto status = conf->GetOptionString(opts, &result);
+        *value = "{" + result + "}";
+        return status;
+      },
+      [](const ConfigOptions& opts, const std::string& name, const char* addr1,
+         const char* addr2, std::string* mismatch) {
+        const auto this_one =
+            reinterpret_cast<const ColumnFamilyOptions*>(addr1);
+        const auto that_one =
+            reinterpret_cast<const ColumnFamilyOptions*>(addr2);
+        auto this_conf = CFOptionsAsConfigurable(*this_one);
+        auto that_conf = CFOptionsAsConfigurable(*that_one);
+        std::string mismatch_opt;
+        bool result =
+            this_conf->AreEquivalent(opts, that_conf.get(), &mismatch_opt);
+        if (!result) {
+          *mismatch = name + "." + mismatch_opt;
+        }
+        return result;
+      }}},
+};
+
+static std::unordered_map<std::string, OptionTypeInfo> cs_input_type_info = {
+    {"column_family", OptionTypeInfo::Struct("column_family", &cfd_type_info, 0,
```

Comment:
The 0 should probably be offset_of, right?  If someone changes the struct, won't this break?

Reply:
sure, I'm just following the other existing structure definition.